edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs

```diff
@@ -0,0 +1,324 @@
+            if (! item.Metadata.Labels.ContainsKey(Constants.k8sEdgeModuleLabel))
+            {
+                return;
+            }
+
+            string podName = item.Metadata.Labels[Constants.k8sEdgeModuleLabel];
+            Events.PodStatus(type, podName);
+            switch (type)
+            {
+                case WatchEventType.Added:
+                case WatchEventType.Modified:
+                case WatchEventType.Error:
+                    ModuleRuntimeInfo runtimeInfo = this.ConvertPodToRuntime(podName, item);
+                    using (await this.moduleLock.LockAsync())
+                    {
+                        this.moduleRuntimeInfos[podName] = runtimeInfo;
+                        this.OnModulesChanged();
+                    }
+
+                    break;
+                case WatchEventType.Deleted:
+                    using (await this.moduleLock.LockAsync())
+                    {
+                        ModuleRuntimeInfo removedRuntimeInfo;
+                        if (!this.moduleRuntimeInfos.TryRemove(podName, out removedRuntimeInfo))
+                        {
+                            Events.PodStatusRemoveError(podName);
+                        }
+
+                        this.OnModulesChanged();
+
+                    }
+
+                    break;
+
+            }
+        }
+
+        (ModuleStatus, string) ConvertPodStatusToModuleStatus(Option<V1ContainerStatus> podStatus)
+        {
+            // TODO: Possibly refine this?
+            return podStatus.Map(
+                pod =>
+                {
+                    if (pod.State != null)
+                    {
+                        if (pod.State.Running != null)
+                        {
+                            return (ModuleStatus.Running, $"Started at {pod.State.Running.StartedAt.GetValueOrDefault(DateTime.Now)}");
+                        }
+
+                        if (pod.State.Terminated != null)
+                        {
+                            return (ModuleStatus.Failed, pod.State.Terminated.Message);
+                        }
+
+                        if (pod.State.Waiting != null)
+                        {
+                            return (ModuleStatus.Failed, pod.State.Waiting.Message);
+                        }
+                    }
+
+                    return (ModuleStatus.Unknown, "Unknown");
+                }).GetOrElse(() => (ModuleStatus.Unknown, "Unknown"));
+        }
+
+        static Option<V1ContainerStatus> GetContainerByName(string name, V1Pod pod)
+        {
+            string containerName = KubeUtils.SanitizeDNSValue(name);
+            return pod.Status?.ContainerStatuses
+                .Where(status => string.Equals(status.Name, containerName, StringComparison.OrdinalIgnoreCase))
+                .Select(status => Option.Some(status))
+                .FirstOrDefault() ?? Option.None<V1ContainerStatus>();
```

Comment:
If it's default, will that be null? I ask because if it's an empty List then it will not trigger the Option.None

Reply:
The FirstOrDefault should return default(Option<V1ContainerStatus>), which is Option.None<V1ContainerStatus>.